tests/cpptests/test_cpp_iterator_union.cpp

```diff
@@ -0,0 +1,268 @@
+/*
+ * Copyright Redis Ltd. 2016 - present
+ * Licensed under your choice of the Redis Source Available License 2.0 (RSALv2) or
+ * the Server Side Public License v1 (SSPLv1).
+ */
+
+#include "rmutil/alloc.h"
+
+#include "gtest/gtest.h"
+#include "iterator_util.h"
+
+#include "src/iterators/union_iterator.h"
+
+class UnionIteratorCommonTest : public ::testing::TestWithParam<std::tuple<unsigned, bool, std::vector<t_docId>>> {
+protected:
+    std::vector<std::vector<t_docId>> docIds;
+    std::vector<t_docId> resultSet;
+    QueryIterator *ui_base;
+
+    void SetUp() override {
+        ASSERT_EQ(RSGlobalConfig.iteratorsConfigParams.minUnionIterHeap, 20) <<
+            "If we ever change the default threshold for using heaps, we need to modify the tests "
+            "here so they still check both flat and heap alternatives.";
+
+        auto [numChildren, quickExit, union_res] = GetParam();
+        // Set resultSet to the expected union result
+        resultSet = union_res;
+        // Set docIds so the union of all children is union_res.
+        // Make sure that some ids are repeated in some children
+        docIds.resize(numChildren);
+        for (size_t i = 0; i < union_res.size(); i++) {
+            for (unsigned j = 0; j < numChildren; j++) {
+                if (j % (i + 1) == 0) {
+                    docIds[j].push_back(union_res[i]);
+                }
+            }
+        }
+        // Create children iterators
+        auto children = (QueryIterator **)rm_malloc(sizeof(QueryIterator *) * numChildren);
+        for (unsigned i = 0; i < numChildren; i++) {
+            children[i] = (QueryIterator *) new MockIterator(docIds[i]);
+        }
+        // Create a union iterator
+        ui_base = IT_V2(NewUnionIterator)(children, numChildren, quickExit, 1.0, QN_UNION, NULL, &RSGlobalConfig.iteratorsConfigParams);
+    }
+    void TearDown() override {
+        ui_base->Free(ui_base);
+    }
+};
+
+TEST_P(UnionIteratorCommonTest, Read) {
+    UnionIterator *ui = (UnionIterator *)ui_base;
+    IteratorStatus rc;
+
+    // Test reading until EOF
+    size_t i = 0;
+    while ((rc = ui_base->Read(ui_base)) == ITERATOR_OK) {
```

Comment:
does this test make sure that in some scenarios the mock iterator returns `ITERATOR_NOTFOUND`?

Reply:
We have cases where the union itself returns `ITERATOR_NOTFOUND`, which means all the children returned it as well. We test all the possible ids in the range [1..maxId] (`maxId` in the expected result set) so we must have scenarios where some iterators found the ids and some don't (also we have cases where all the children return `OK` as well)